cloud/common.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	return responseBody, nil
+}
+
+// PostRaw sends a POST with raw data and returns the raw response body
+// Errors from the API are returned as an error
+func (c Client) PostRaw(path, body string, headers map[string]string) ([]byte, error) {
+	resp, err := c.Request("POST", path, body, headers)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pass the body to the callback function
+	c.rawFunc("POST", path, resp.StatusCode, responseBody)
+
+	// TODO make sure this covers cloud API errors...
+	// Convert the API's JSON error response to an error
+	if resp.StatusCode == 400 || resp.StatusCode == 404 || resp.StatusCode == 500 {
+		return responseBody, fmt.Errorf("POST %s failed with status %d: %s", path, resp.StatusCode, ErrorToString(responseBody))
+	}
+
+	return responseBody, nil
+}
+
+// PostJSON sends a POST with JSON data and the Content-Type header set to "application/json"
+// Errors from the API are returned as an error
+func (c Client) PostJSON(path, body string) ([]byte, error) {
+	headers := map[string]string{"Content-Type": "application/json"}
+	return c.PostRaw(path, body, headers)
+}
+
+// IsStringInSlice returns true if the string is present, false if not
+// slice must be sorted
+func IsStringInSlice(slice []string, s string) bool {
+	i := sort.SearchStrings(slice, s)
+	if i < len(slice) && slice[i] == s {
+		return true
+	}
+	return false
+}
+
+// GetContentFilename returns the filename from a content disposition header
+func GetContentFilename(header string) (string, error) {
+	// Get the filename from the content-disposition header
+	// Split it on ; and strip whitespace
+	parts := strings.Split(header, ";")
+	for _, p := range parts {
+		p = strings.TrimSpace(p)
+		fields := strings.Split(p, "=")
+		if len(fields) == 2 && fields[0] == "filename" {
+			filename := filepath.Base(strings.TrimSpace(fields[1]))
+
+			if filename == "/" || filename == "." || filename == ".." {
+				return "", fmt.Errorf("Invalid filename in header: %s", p)
+			}
+			return filename, nil
+		}
+	}
+	return "", fmt.Errorf("No filename in header: %s", header)
+}
+
+// MoveFile will copy the src file to the destination file and remove the source on success
+// It assumes the destination file doesn't exist, or if it does that it should be overwritten
+func MoveFile(src, dst string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+	_, err = io.Copy(dstFile, srcFile)
+	if err == nil {
+		srcFile.Close()
+		os.Remove(src)
+	}
+	return err
+}
+
+// AppendQuery adds the query string to the current url using ? for the first and & for subsequent ones
+func AppendQuery(url, query string) string {
```

Comment:
This can also be done through `net/url`.

Reply:
True. See comment on other url handling.